### DISCUSSION TITLE
API throttling

### DISCUSSION THREAD

**Discussion by ipaleka** [asastats/channel#230] [General]
ASA Stats website users' address pages are cached for 59 seconds. No more than 1 (unique) address response is available per minute for free.

This thread should define throttling and a number of requests per minute for the API users - in this phase of the project that implies DAO governors' limits who are only legit API users as of today.

Jay's ASA Stats mobile app and @dragmz terminal CLI app are expected to follow the same limit as the website.

DAO governors are invited to ping Discord admins and get access to #mobile-testing Discord channel, they can get a link for the terminal app access the same way.

### Background

As defined in our [utilization document](https://github.com/asastats/channel/discussions/166), every subscriber starting from the Asastatser tier gets access to the ASA Stats API.  

ASA Stats mobile (iOS and Android) is an official ASA Stats app and its usage of the API follows global ASA Stats permissions for free users (1 uncached request per minute).

@dragmz terminal app can be used in two different ways: 

a) hosted on @dragmz server and by using his API credentials and bandwidth
   It's a DragMZ decision on how he'd allocate his available resources to the users of his app.

b) as a standalone version using your API credentials and resources
   Your bandwidth and throttling are defined by the size of your governor seat (number of tokens).

**Comment by ipaleka** [comment]
By hanniballlz in [Discord](https://discord.com/channels/906917846754418770/1019690031918612480/1027322837645672468):

Is the API available now? I think it would be useful in the project I'm currently working on but from the website it seems like it isn't published yet?

- - - - -

ipaleka — Today at 11:11 AM
Right now only ASA Stats DAO governors are qualified to receive their API credentials. We'll surely discuss the possible ways for others to be included in this early phase before our Permission ASC is deployed and most of the extended functionalities are available for subscribers.

One of the options is to stake ASASTATS in a dedicated utilization staking program (Cometa or some other provider, still to discuss...) or provide liquidity to Tinyman's ASASTATS-ALGO LP for a period of one year. From the day one you'll get permissions and access to the ASA Stats API based on ASASTATS involved in the process, while after that year passes you get permanent governance seat and a dedicated number of governance votes and related level in the Permission ASC.

May you explain the nature of your project and some specifics you may need from our API?
- - - - -

**Reply by hannalwilson** [reply]
I am a developer for Alchemon and we are finalizing our governance voting. Our governance token, AlcheCoin, can be staked in liquidity pools and we want to include those tokens in users governance votes. So we need a way to see how much ALCH users have staked and someone mentioned ASA Stats to me so I checked it out and saw your site displays the total ALCH I hold, including what is in the LP. Essentially  all I need from the API is the amount of ALCH specific users hold in LP.

**Reply by ipaleka** [reply]
> I am a developer for Alchemon and we are finalizing our governance voting. Our governance token, AlcheCoin, can be staked in liquidity pools and we want to include those tokens in users governance votes. So we need a way to see how much ALCH users have staked and someone mentioned ASA Stats to me so I checked it out and saw your site displays the total ALCH I hold, including what is in the LP. Essentially all I need from the API is the amount of ALCH specific users hold in LP.

We have been reached by another source that covers pretty much the same scenario, so we'll use this subthread you just created as a discussion for the [following issue](https://github.com/asastats/channel/issues/270).

**Reply by ipaleka** [reply]
> I am a developer for Alchemon and we are finalizing our governance voting. Our governance token, AlcheCoin, can be staked in liquidity pools and we want to include those tokens in users governance votes.

May I ask about the exact process, will you be committing ALGO to the Foundation app on behalf of your users, and then disqualifying them with your own eligibility-checking mechanism? Or you'll be providing data to the Foundation?

> Essentially all I need from the API is the amount of ALCH specific users hold in LP.

Do you need the amount of ALCH farmed in the ALGO pools, or ALCH farmed in any pool?

Or to ask the right question: is this for your own governance or for the Algorand Governance?

P.S. Or to answer my own question: https://alchemon.net/governance
> AlcheCoin is Alchemon's Governance Token. It gives holders the right to vote on the development and operations of Alchemon, i.e., it gives our community decision making power. Each AlcheCoin is worth one vote and each wallet is only allowed to vote once. Before voting opens, a snapshot of all AlcheCoin is taken and the amount of AlcheCoin held in your wallet at the time of the snapshot is how many times your vote counts. For example, if you had 5400 AlcheCoin and voted for option A, that would count as 5400 'A' votes.

**Reply by ipaleka** [reply]
Here's a screenshot of ALCH section from a random address API response:
![alch](https://user-images.githubusercontent.com/49662536/194378297-d3bd8067-cda7-4216-bd15-8878f87ba8b6.png)

<details> 
  <summary>Here's an extracted ALCH part of a random address API response</summary>
    {
      "value": 1437.037720288807,
      "id": 310014962,
      "amount": 111566,
      "asa": {
        "id": 310014962,
        "name": "AlcheCoin",
        "unit": "ALCH",
        "total": 100000000,
        "decimals": 0,
        "url": "",
        "links": [{
          "provider": "vestige",
          "link": "https://vestige.fi/asset/310014962/",
          "title": "ALCH-ALGO Vestige page"
        }, {
          "provider": "algocharts",
          "link": "https://algocharts.net/chart1m.php?asset_in={asset_id}&asset_out=0",
          "title": "ALCH-ALGO AlgoCharts page"
        }]
      },
      "dapps": [{
        "dapp": "hp",
        "entries": {
          "value": {
            "linktext": null,
            "link": null,
            "title": null,
            "text": null,
            "border": null,
            "value": 922.4421731960373,
            "bracket": null
          },
          "links": [{
            "small": null,
            "linktext": "Humble LP ALGO-ALCH",
            "link": "https://app.humble.sh/pool",
            "title": "Go to Humble Swap liquidity pool"
          }, {
            "small": null,
            "linktext": "Humble Farm",
            "link": "https://app.humble.sh/farm",
            "title": "Go to Humble Farm"
          }],
          "amount": {
            "text": null,
            "amount": 71729,
            "unit": "ALCH",
            "border": null,
            "bracket": null
          },
          "price": {}
        },
        "proxy": null,
        "distribution": [{
          "value": 824.6667284085083,
          "amount": 64126.0,
          "price": 0.012860099310864678,
          "text": "Humble swap",
          "link": "https://app.humble.sh/swap?poolId=779144473&asset_in=310014962&asset_out=0"
        }, {
          "value": 97.775444787529,
          "amount": 7603.0,
          "price": 0.012860113742934236,
          "text": "Tinyman swap",
          "link": "https://app.tinyman.org/#/swap?asset_in=310014962&asset_out=0"
        }]
      }, {
        "dapp": "e",
        "entries": {
          "value": {
            "linktext": null,
            "link": null,
            "title": null,
            "text": null,
            "border": null,
            "value": 514.5955470927696,
            "bracket": null
          },
          "links": [],
          "amount": {
            "text": "Balance:",
            "amount": 39837,
            "unit": "ALCH",
            "border": true,
            "bracket": null
          },
          "price": {
            "price": 0.012917527602298607,
            "unit": "ALGO/ALCH"
          }
        },
        "proxy": null,
        "distribution": [{
          "value": 498.4485271765442,
          "amount": 38587.0,
          "price": 0.01291752474088538,
          "text": "Humble swap",
          "link": "https://app.humble.sh/swap?poolId=779144473&asset_in=310014962&asset_out=0"
        }, {
          "value": 16.14701991622543,
          "amount": 1250.0,
          "price": 0.012917615932980346,
          "text": "Tinyman swap",
          "link": "https://app.tinyman.org/#/swap?asset_in=310014962&asset_out=0"
        }]
      }],
      "price": 0.013005266250610352
    }
</details>

**Reply by hannalwilson** [reply]
Yeah, we are doing this for our own governance, not Algorands. I appreciate you taking the time to discuss this with me! I was able to find a publicly available API that fulfilled our needs for governance. However, I really like what ASA Stats is and the direction it is going so I would definitely be interested in becoming a DAO Governor, if that is possible.

**Reply by ipaleka** [reply]
> Yeah, we are doing this for our own governance, not Algorands. I appreciate you taking the time to discuss this with me! I was able to find a publicly available API that fulfilled our needs for governance. However, I really like what ASA Stats is and the direction it is going so I would definitely be interested in becoming a DAO Governor, if that is possible.

Well, we provide governance seats to the contributors coming from our/ASA Stats community, as well as to the Algorand community members who help us with implementing their projects in ASA Stats.

The latest/third way is by purchasing our [subscription tiers](https://github.com/asastats/channel/discussions/166), but the official deployment of our Permission ASC is not near around. That being said, we're looking for other ways for people and projects to use our API before the system based on Permission ASC takes place. And dedicated ASASTATS staking or added liquidity to Tinyman's ASASTATS-ALGO LP are two possible/probable options as of now.

Meanwhile, if we get closer to a solution, we'd supply testing API credentials to you so you can use them during your development.
